src/admin_ws.rs

```diff
@@ -0,0 +1,187 @@
+
+    /// Checks if there is an active connection
+    fn is_connected(&self) -> HcHttpGatewayResult<bool> {
+        let connection = self.connection_handle.lock().map_err(|e| {
+            HcHttpGatewayError::InternalError(format!(
+                "Mutex was poisoned during is_connected check: {}",
+                e
+            ))
+        })?;
+
+        Ok(connection.is_some())
+    }
+
+    /// Gets reconnection retry count
+    pub fn get_reconnection_retries(&self) -> usize {
+        self.current_retries
+    }
+
+    /// Ensures that a connection is established before proceeding
+    async fn ensure_connected(&mut self) -> HcHttpGatewayResult<()> {
+        if self.is_connected()? {
+            return Ok(());
+        }
+
+        self.connect().await
+    }
+
+    /// Attempts to connect to the AdminWebsocket
+    ///
+    /// This will make multiple attempts according to the `max_retries` and
+    /// `retry_delay_ms` settings, with exponential backoff between retries.
+    pub async fn connect(&mut self) -> HcHttpGatewayResult<()> {
+        self.current_retries = 0;
+
+        while self.current_retries < ADMIN_WS_CONNECTION_MAX_RETRIES {
+            match AdminWebsocket::connect(&self.url).await {
+                Ok(conn) => {
+                    let mut connection = self.connection_handle.lock().map_err(|e| {
+                        HcHttpGatewayError::InternalError(format!(
+                            "Mutex was poisoned during is_connected check: {}",
+                            e
+                        ))
+                    })?;
+
+                    *connection = Some(conn);
+                    self.current_retries = 0;
+
+                    return Ok(());
+                }
+                Err(e) => {
+                    self.current_retries += 1;
+                    tracing::warn!(
+                        "Failed to connect to WebSocket (attempt {}/{}): {:?}",
+                        self.current_retries,
+                        ADMIN_WS_CONNECTION_MAX_RETRIES,
+                        e
+                    );
+
+                    if self.current_retries < ADMIN_WS_CONNECTION_MAX_RETRIES {
+                        sleep(Duration::from_millis(ADMIN_WS_CONNECTION_RETRY_DELAY_MS)).await;
+                    } else {
+                        return Err(HcHttpGatewayError::UpstreamUnavailable);
+                    }
+                }
+            }
+        }
+
+        Err(HcHttpGatewayError::ConfigurationError(format!(
+            "Maximum connection retry attempts ({}) reached",
+            ADMIN_WS_CONNECTION_MAX_RETRIES
+        )))
+    }
+
+    /// Allows calling a method on the AdminWebsocket, with automatic reconnection if needed
+    ///
+    /// Accepts a function `f` that takes a AdminWebsocket and returns a Result with ConductorApiError
+    pub async fn call<T, F, Fut>(&mut self, f: F) -> HcHttpGatewayResult<T>
+    where
+        F: Fn(Arc<AdminWebsocket>) -> Fut + Send + Clone,
```

Comment:
Any reason not to use a `BoxFuture` here? Deals with getting the trait bounds right so it's simpler to work with